spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,412 @@
+            elif s == 2:
+                short_form = sample_toks[0] + os.sep + sample_toks[1]
+            else:
+                short_form = "..." + os.sep + sample_toks[s-1]
+            for idx in level_idx:
+                new_path_list[idx] += short_form + os.sep
+                level_idx[idx] = level_idx[idx][s:]
+
+        # Group the remaining bit after the common prefix, shorten, and recurse
+        while level_idx:
+            k, group = 0, level_idx  # k is length of the group's common prefix
+            while True:
+                # Abort if we've gone beyond end of one or more in the group
+                prospective_group = {idx: toks for idx, toks
+                                     in group.items() if len(toks) == k}
+                if prospective_group:
+                    if k == 0:  # we spit out the group with no suffix
+                        group = prospective_group
+                    break
+                # Only keep going if all n still match on the kth token
+                _, sample_toks = next(group.iteritems())
+                prospective_group = {idx: toks for idx, toks
+                                     in group.items()
+                                     if toks[k] == sample_toks[k]}
+                if len(prospective_group) == len(group) or k == 0:
+                    group = prospective_group
+                    k += 1
+                else:
+                    break
+            _, sample_toks = next(group.iteritems())
+            if k == 0:
+                short_form = ''
+            elif k == 1:
+                short_form = sample_toks[0]
+            elif k == 2:
+                short_form = sample_toks[0] + os.sep + sample_toks[1]
+            else:  # k > 2
+                short_form = sample_toks[0] + "..." + os.sep + sample_toks[k-1]
+            for idx in group.keys():
+                new_path_list[idx] += short_form + (os.sep if k > 0 else '')
+                del level_idx[idx]
+            recurse_level({idx: toks[k:] for idx, toks in group.items()})
+
+    recurse_level({i: pl for i, pl in enumerate(path_list) if pl})
+    return [path.rstrip(os.sep) for path in new_path_list]
+
+
+def BuildFuzzyFormatter(needle):
+    """Matches on the first occurance of the letters in needle in order,
+    wrapping those letters with <b></b> tags. If no match is found None
+    is returend.
+    """
+    re_exons = re.compile(''.join("([^" + re.escape(c) + "]*)" + re.escape(c)
+                                  for c in needle) + '(.*)')
+
+    def func(text):
+        exons = re_exons.match(text)
+        if not exons:
+            return
+        exons = exons.groups()
+        ret = ''.join(e + i for e, i in
+                      zip(exons, ['<b>' + c + '</b>'for c in needle]))
+        return ret + exons[-1]
+    return func
+
+
+class HTMLDelegate(QStyledItemDelegate):
+    """With this delegate, a QListWidgetItem can render HTML.
```

Comment:
@ccordoba12 I am using the exact same delegate for the Keysequence shortcut editor (to show the rich text), should we also add this to the helper widgets files as well? or should it go to another file with a different name (like delegates...)